src/v/archival/segment_reupload.cc

```diff
@@ -175,21 +188,21 @@ void segment_collector::align_end_offset_to_manifest(
 
 segment_collector::lookup_result segment_collector::find_next_segment(
   model::offset start_offset, segment_collector_mode mode) {
-    const auto& segment_set = _log.segments();
-    auto it = segment_set.lower_bound(start_offset);
-    // start_offset < log start due to eviction of segments before
-    // they could be uploaded, skip forward to log start.
+    // 'start_offset' should always be above the start offset of the local log
+    // as we skip to it in the calling code (`collect_segments`).
     if (start_offset < _log.offsets().start_offset) {
         vlog(
-          archival_log.debug,
-          "Finding next segment for {}: start_offset: {} behind log "
-          "start: {}, skipping forward",
+          archival_log.warn,
+          "Finding next segment for {}: can't find segments below the local "
+          "log start offset ({} < {})",
           _manifest.get_ntp(),
           start_offset,
           _log.offsets().start_offset);
-        it = segment_set.begin();
+        return {};
```

Comment:
Nice find, I do have a question about the changes here.

If there is a log eviction between setting `_begin_inclusive` in `collect_segments` and this point (line 193), would we end up collecting no compacted segments? In that case we should skip forward, right? 

There are no scheduling points between the two calls so the possibility of that happening does seem pretty low (or non existent, if the same shard runs the archiver and the disk log impl?).

Reply:
> If there is a log eviction between setting _begin_inclusive in collect_segments and this point (line 193), would we end up collecting no compacted segments?

Yes, that's right.

> There are no scheduling points between the two calls so the possibility of that happening does seem pretty low (or non existent, if the same shard runs the archiver and the disk log impl?).

Indeed, the change assumes no scheduling points between setting `_begin_inclusive` in `collect_segments` and this point (line 193). The bug would still be fixed, but we would not re-upload anything in that case. I think this is fine though as the same shard should always run both the archiver and the disk log.

Reply:
Ok, the change lgtm in that case